_posts/2019/2019-12-25-adding-hateoas-to-an-aspnetcore-rest-api.md

```diff
@@ -0,0 +1,297 @@
+
+    clientModel.Links.Add(
+        UrlLink("all", "GetClients", null));
+
+    clientModel.Links.Add(
+        UrlLink("_self", "GetClientAsync", new { id = clientModel.Id }));
+
+    clientModel.Links.Add(
+        UrlLink("addresses", "GetAddressesByClient", new { id = clientModel.Id }));
+
+    return clientModel;
+}
+
+{% endhighlight %}
+
+In the `RestfulClient` method above, we take in a Client object and convert it to a ClientModel with AutoMapper.  Then we add links for "all", "_self" and "addresses" paths.  
+
+The "all" relationship will provide a link to our API method that returns all clients.  The "_self" relationship defines where this client can be retrieved from the API. The "addresses" relationship defines where to find the addresses associated with this client. Each of the methods we define in those links ('GetClients', 'GetClientAsync' and 'GetAddressesByClient') will be defined later in our ClientController.
+
+### Adding Controllers
+
+Now that our base controller is done, let's add a `ClientController.cs` that inherits from our `RestControllerBase` class. First, we'll define a constructor and inject everything our base class needs as well as our DbContext.
+
+{% highlight csharp %}
+
+private readonly ILogger<ClientController> _logger;
+private readonly BBBContext _bbbContext;
+
+public ClientController(
+    ILogger<ClientController> logger, 
+    IActionDescriptorCollectionProvider actionDescriptorCollectionProvider, 
+    BBBContext bbbContext, 
+    IMapper mapper)
+    : base(actionDescriptorCollectionProvider, mapper)
+{
+    _logger = logger;
+    _bbbContext = bbbContext;
+}
+
+{% endhighlight %}
+
+Now we can access our database to retrieve data and return it using the `RestfulClient` method of our base class.
+
+{% highlight csharp %}
+
+[HttpGet(Name = "GetClients")]
+public IActionResult GetClients()
+{
+    IEnumerable<Client> clients = _bbbContext.Clients;
+    IEnumerable<ClientModel> clientModels = clients
+                                            .Select(f => RestfulClient(f));
+    
+    return Ok(clientModels);
+}
+
+[HttpGet("{id}", Name = "GetClientAsync")]
+public async Task<IActionResult> GetClientAsync(Guid id)
+{
+    if (id == Guid.Empty)
+    {
+        return BadRequest();
+    }
+
+    Client client = await _bbbContext.Clients.FindAsync(id);
+
+    if (client == null)
+    {
+        return NotFound();
+    }
+
+    return Ok(RestfulClient(client));
+}
+
+[HttpGet("{id}/addresses", Name = "GetAddressesByClient")]
+public IActionResult GetAddressesByClient(Guid id)
+{
+    if (id == Guid.Empty)
+    {
+        return BadRequest();
+    }
+
+    IEnumerable<Address> addresses = _bbbContext.Addresses
+                                                .Where(w => 
+                                                    w.ClientId.Equals(id));
+    IEnumerable<AddressModel> addressModels = addresses.Select(f => 
+                                                            RestfulAddress(f));
```

Comment:
This clause gets rendered really weirdly in your CMS - looks like it's not picking up the spaces for some reason